cdap-ui/app/directives/log-viewer/log-viewer.js

```diff
@@ -626,6 +630,60 @@ function LogViewerController($scope, $window, LogViewerStore, myLogsApi, LOGVIEW
     startTimeRequest();
   };
 
+  vm.setHoverLogLevels = function (eventType) {
+    vm.resetHoverLogLevels();
+    logLevelsFallThrough(vm.activeHoverLogLevels, eventType);
+    vm.setActiveTransitionChecks();
+  };
+
+  vm.resetHoverLogLevels = function () {
+    //Identifies what log levels should be in hovered state
+    vm.activeHoverLogLevels = {
+      'ERROR': false,
+      'WARN': false,
+      'INFO': false,
+      'DEBUG': false,
+      'TRACE': false
+    };
+    // Transition state - levels that would be added to active log levels or
+    // levels that would be removed from active log levels i.e greyed out levels
+    vm.activeTransitionChecks = {
+      'ERROR': false,
+      'WARN': false,
+      'INFO': false,
+      'DEBUG': false,
+      'TRACE': false
+    };
+  };
+
+  vm.setActiveTransitionChecks = function () {
+    //This function identifies which log levels are in transition state.
+    const noLevelHovered = !Object.values(vm.activeHoverLogLevels).some(el => el === true);
+    //Iterate over all log levels
+    Object.keys(vm.activeLogLevels).forEach(option => {
+      if (vm.activeLogLevels[option]) {
+        if (vm.activeHoverLogLevels[option]) {
```

Comment:
Since vm.activeTransitionChecks[option] is only true when current level is active but not hovered and there is at least one level in the hovered state, this if block could potentially be simplified to something like

vm.activeTransitionChecks[options] = !vm.activeHoverLogLevels[option] && !noLevelHovered

Reply:
@yukiej Yes, but do you think it'll be easy to understand when reading this code later? @ajainarayanan @elfenheart thoughts?

Reply:
That's a good point. The main motivation of the suggestion was to make it more clear when vm.activeTransitionChecks[options] is set to true, but that does make it less clear what the other possibilities are.

Reply:
I don't have a strong opinion about this. @ajainarayanan ?

Reply:
I think its ok to have this as `if..else` condition. Our build minifier will optimize it for production build. 
👏 👏 great job in writing comments explaining the logic @itsanudeep!